server/src/main/scala/io/delta/standalone/internal/DeltaSharingCDCReader.scala

```diff
@@ -0,0 +1,281 @@
+      latestVersion
+    )
+
+    if (endingVersion.exists(_ < startingVersion.get)) {
+      throw DeltaCDFErrors.endBeforeStartVersionInCDF(startingVersion.get, endingVersion.get)
+    }
+
+    (startingVersion.get, endingVersion.getOrElse(latestVersion))
+  }
+
+  // Convert timestamp string in cdfOptions to Timestamp
+  private def getTimestamp(paramName: String, timeStampStr: String): Timestamp = {
+    try {
+      Timestamp.valueOf(timeStampStr)
+    } catch {
+      case e: IllegalArgumentException =>
+        throw DeltaCDFErrors.invalidTimestamp(paramName, e.getMessage)
+    }
+  }
+
+  /**
+   * - If a commit version exactly matches the provided timestamp, we return it.
+   * - Otherwise, we return the earliest commit version
+   *   with a timestamp greater than the provided one.
+   * - If the provided timestamp is larger than the timestamp
+   *   of any committed version, we throw an error.
+   *
+   * @param timestamp - user specified timestamp
+   * @param latestVersion - latest version of the table
+   * @return - corresponding version number for timestamp
+   */
+  private def getStartingVersionFromTimestamp(timestamp: Timestamp, latestVersion: Long): Long = {
+    val commit = history.getActiveCommitAtTime(timestamp)
+    if (commit.timestamp >= timestamp.getTime) {
+      // Find the commit at the `timestamp` or the earliest commit
+      commit.version
+    } else {
+      // commit.timestamp is not the same, so this commit is a commit before the timestamp and
+      // the next version if exists should be the earliest commit after the timestamp.
+      // Note: `getActiveCommitAtTime` has called `update`, so we don't need to call it again.
+      if (commit.version + 1 <= latestVersion) {
+        commit.version + 1
+      } else {
+        val commitTs = new Timestamp(commit.timestamp)
+        throw DeltaErrors.timestampLaterThanTableLastCommit(timestamp, commitTs)
+      }
+    }
+  }
+
+  /**
+   * Given timestamp or version this method returns the corresponding version for that timestamp
+   * or the version itself.
+   */
+  private def getVersionForCDC(
+      options: Map[String, String],
+      versionKey: String,
+      timestampKey: String,
+      latestVersion: Long
+  ): Option[Long] = {
+    if (options.contains(versionKey)) {
+      Some(options(versionKey).toLong)
+    } else if (options.contains(timestampKey)) {
+      val tsStr = options(timestampKey)
+      if (timestampKey == DeltaDataSource.CDF_START_TIMESTAMP_KEY) {
+        // For the starting timestamp we need to find a version after the provided timestamp
+        // we can use the same semantics as streaming.
+        val ts = getTimestamp("startingTimestamp", tsStr)
+        val resolvedVersion = getStartingVersionFromTimestamp(ts, latestVersion)
+        Some(resolvedVersion)
+      } else {
+        // For ending timestamp the version should be before the provided timestamp.
+        val ts = getTimestamp("endingTimestamp", tsStr)
```

Comment:
"endTimestamp" -> timestampKey?

could we move the ts = ... assignment out of the if/else ?

Reply:
updated.